File: lib/py/uapi/internal/schema/DerivePossibleSelects.py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from uapi.internal.types.UUnion import UUnion
    from uapi.internal.types.UStruct import UStruct
    from uapi.internal.types.UType import UType
    from uapi.internal.types.UFieldDeclaration import UFieldDeclaration


def derive_possible_select(fn_name: str, result: 'UUnion') -> dict[str, object]:
    from uapi.internal.types.UUnion import UUnion
    from uapi.internal.types.UStruct import UStruct

    nested_types: dict[str, UType] = {}
    ok_fields: dict[str, UFieldDeclaration] = result.cases['Ok_'].fields

    ok_field_names = sorted(ok_fields.keys())

    find_nested_types(ok_fields, nested_types)

    possible_select: dict[str, object] = {}

    possible_select['->'] = {
        'Ok_': ok_field_names,
    }

    sorted_type_keys = sorted(nested_types.keys())
    for k in sorted_type_keys:
        print(f'k: {k}')
        v = nested_types[k]
        if isinstance(v, UUnion):
            union_select: dict[str, list[str]] = {}
            sorted_case_keys = sorted(v.cases.keys())
            for c in sorted_case_keys:
                typ = v.cases[c]
                selected_field_names: list[str] = sorted(typ.fields.keys())
                union_select[c] = selected_field_names

            possible_select[k] = union_select
        elif isinstance(v, UStruct):
            struct_select: list[str] = sorted(v.fields.keys())
            possible_select[k] = struct_select

    return possible_select


def find_nested_types(fields: dict[str, 'UFieldDeclaration'], nested_types: dict[str, 'UType']) -> None:
    from uapi.internal.types.UUnion import UUnion
    from uapi.internal.types.UStruct import UStruct

    for field in fields.values():
        typ = field.type_declaration.type
        if isinstance(typ, UUnion):
            nested_types[typ.name] = typ
            for c in typ.cases.values():
                find_nested_types(c.fields, nested_types)
        elif isinstance(typ, UStruct):
            nested_types[typ.name] = typ
            find_nested_types(typ.fields, nested_types)

File: lib/java/src/main/java/uapi/internal/schema/DerivePossibleSelects.java
package uapi.internal.schema;

import java.util.*;

import uapi.internal.types.UFieldDeclaration;
import uapi.internal.types.UStruct;
import uapi.internal.types.UType;
import uapi.internal.types.UUnion;

public class DerivePossibleSelects {

    public static Map<String, Object> derivePossibleSelect(String fnName, UUnion result) {
        final var nestedTypes = new HashMap<String, UType>();
        final var okFields = result.cases.get("Ok_").fields;

        final var okFieldNames = new ArrayList<>(okFields.keySet());
        Collections.sort(okFieldNames);

        findNestedTypes(okFields, nestedTypes);

        final var possibleSelect = new HashMap<String, Object>();

        possibleSelect.put("->", Collections.singletonMap("Ok_", okFieldNames));

        final var sortedTypeKeys = new ArrayList<>(nestedTypes.keySet());
        Collections.sort(sortedTypeKeys);
        for (final var k : sortedTypeKeys) {
            System.out.println("k: " + k);
            final var v = nestedTypes.get(k);
            if (v instanceof UUnion u) {
                final var unionSelect = new HashMap<String, List<String>>();
                final var sortedCaseKeys = new ArrayList<>(u.cases.keySet());
                Collections.sort(sortedCaseKeys);
                for (final var c : sortedCaseKeys) {
                    final var typ = u.cases.get(c);
                    final var selectedFieldNames = new ArrayList<String>();
                    final var sortedFieldNames = new ArrayList<>(typ.fields.keySet());
                    Collections.sort(sortedFieldNames);
                    selectedFieldNames.addAll(sortedFieldNames);

                    unionSelect.put(c, selectedFieldNames);
                }

                possibleSelect.put(k, unionSelect);
            } else if (v instanceof UStruct s) {
                final var structSelect = new ArrayList<String>();
                final var sortedFieldNames = new ArrayList<>(s.fields.keySet());
                Collections.sort(sortedFieldNames);
                structSelect.addAll(sortedFieldNames);

                possibleSelect.put(k, structSelect);
            }
        }

        return possibleSelect;
    }

    private static void findNestedTypes(Map<String, UFieldDeclaration> fields, Map<String, UType> nestedTypes) {
        for (final var field : fields.values()) {
            final var typ = field.typeDeclaration.type;
            if (typ instanceof UUnion u) {
                nestedTypes.put(u.name, typ);
                for (final var c : u.cases.values()) {
                    findNestedTypes(c.fields, nestedTypes);
                }
            } else if (typ instanceof UStruct s) {
                nestedTypes.put(s.name, typ);
                findNestedTypes(s.fields, nestedTypes);
            }
        }
    }
}
